### DISCUSSION TITLE
x-forwarded-client-cert header "dropped" in some pods with mTLS

### DISCUSSION THREAD

**Discussion by wmashal** [istio/istio#49513] [Q&A] [answered]
Hi

1- I have a sleep pod under sleep namespace with PeerAuthentication Strict , from this pod I am sending curl requests to Echo Server, httbin and 3thd POD (from our cluster called resource-manager) to test (all pods namespaces PeerAuthentication Strict and all namespaces inject-sidecar enabled)
2- In Echo Server and httpbin response result I can see the spiffe in the x-forward-client-cert header like the following
**"x-forwarded-client-cert": "By=spiffe://cluster.local/ns/echo/sa/default;Hash=675f96c585fe22a517e96551cf174fe53cdb39d41fcb8988df5e08b76c393f40;Subject=\"\";URI=spiffe://cluster.local/ns/sleep/sa/sleep"**
3- In my resource-manager pod I want to use the **spiffe://cluster.local/ns/sleep/sa/sleep** as Identity to differentiate among clients
4- BUT in my POD the x-forwarded-client-cert its not added with headers
5- So I dont know why echo and httpbin received it  and I can see the header ? but resource-manager POD no ?
6- What I know this should be added by default from Istio, no?

all what I need in the x-forward-client-cert header the **spiffe URI spiffe://cluster.local/ns/<namespace>/sa/<service account>**
But x-forward-client-cert dropped some how from the request but I think the header and spiffe should be added automatically by istio mTls, no ?

![image](https://github.com/envoyproxy/envoy/assets/11045408/b5f35ad6-a8e9-45cd-9371-a1241cd15140)
![image](https://github.com/envoyproxy/envoy/assets/11045408/e58aad6a-33f7-45ad-a9f2-dd5b787c1553)
![image](https://github.com/envoyproxy/envoy/assets/11045408/08a88224-13b2-4cd3-93b0-4778d6509cdb)

**Echo result without any envoy filters:**

```
100 10966  100 10966    0     0   760k      0 --:--:-- --:--:-- --:--:--  823k
* Connection #0 to host echoserver.resource-manager left intact
{
  "host": {
    "hostname": "echoserver.resource-manager",
    "ip": "::ffff:127.0.0.6",
    "ips": []
  },
  "http": {
    "method": "GET",
    "baseUrl": "",
    "originalUrl": "/",
    "protocol": "http"
  },
  "request": {
    "params": {
      "0": "/"
    },
    "query": {},
    "cookies": {},
    "body": {},
    "headers": {
      "host": "echoserver.resource-manager",
      "user-agent": "curl/8.6.0",
      "accept": "*/*",
      "x-forwarded-for": "10.96.0.33,10.96.0.33",
      "x-forwarded-proto": "http",
      "x-request-id": "7c543716-e8fb-96eb-8a81-7f3be7d81b6a",
      "x-envoy-attempt-count": "1",
      "x-envoy-external-address": "10.96.0.33",
      "x-forwarded-client-cert": "By=spiffe://cluster.local/ns/resource-manager/sa/default;Hash=335231a63e4030ee41a0ec4e2ca22f920008338f389c290f564dc378208a5c48;Subject=\"\";URI=spiffe://cluster.local/ns/sleep/sa/sleep",
      "x-b3-traceid": "c18c9b9b386a761f628efde80c635bf2",
      "x-b3-spanid": "9282312e9abaad53",
      "x-b3-parentspanid": "628efde80c635bf2",
      "x-b3-sampled": "1"
```

**My Pod response without curl -H , I got error in my logs**
```
{"level":"error","ts":"2024-02-21T09:21:29Z","logger":"SPIFFEAuthProvider","msg":"missing header","correlation_id":"ec6a8051-d299-4b3d-b944-bc15a97d8d81","requestID":"ed │
│ 4466cd-ae2f-46e5-b584-f4206e0f5f4b","error":"couldn't find header: x-forwarded-client-cert",
```
**If I add the header -H with curl for my pod request** 
`get pod -l app=sleep -n sleep -o jsonpath={.items..metadata.name})" -c sleep -n sleep -- curl http://resource-manager-api2.resource-manager/api/resources/system.resource.api.sap/v1/secrets/test/.ua-test-master-secret -v  -H "x-forwarded-client-cert: By=spiffe://cluster.local/ns/echo/sa/default;Hash=675f96c585fe22a517e96551cf174fe53cdb39d41fcb8988df5e08b76c393f40;Subject=\"\";URI=spiffe://cluster.local/ns/sleep/sa/sleep" | jq .`
**I got the right response** 
```
 % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0* Host resource-manager-api2.resource-manager:80 was resolved.
* IPv6: (none)
* IPv4: 10.147.95.238
*   Trying 10.147.95.238:80...
* Connected to resource-manager-api2.resource-manager (10.147.95.238) port 80
> GET /api/resources/system.resource.api.sap/v1/secrets/test/.ua-test-master-secret HTTP/1.1
> Host: resource-manager-api2.resource-manager
> User-Agent: curl/8.6.0
> Accept: */*
> x-forwarded-client-cert: By=spiffe://cluster.local/ns/echo/sa/default;Hash=675f96c585fe22a517e96551cf174fe53cdb39d41fcb8988df5e08b76c393f40;Subject="";URI=spiffe://cluster.local/ns/sleep/sa/sleep
>
< HTTP/1.1 200 OK
< Date: Wed, 21 Feb 2024 09:23:43 GMT
< Content-Type: application/json
< Content-Length: 9290
< Cache-Control: no-store
< Strict-Transport-Security: max-age=31536000; includeSubDomains
< Correlationid: 40496671-efa9-4e6d-927d-88247fe65d0b
< X-Ratelimittotal-Limit-S: 1000
< X-Ratelimit-Limit-S: 500
< X-Ratelimit-Remaining-S: 999
< X-Ratelimit-Reset-S: 1708507424
< X-Ratelimittotal-Limit-Qh: 10000000
< X-Ratelimit-Limit-Qh: 9090909
< X-Ratelimit-Remaining-Qh: 9999999
< X-Ratelimit-Reset-Qh: 1708508323
<
{ [9290 bytes data]
100  9290  100  9290    0     0  71094      0 --:--:-- --:--:-- --:--:-- 70916
* Connection #0 to host resource-manager-api2.resource-manager left intact
{
  "apiVersion": "system.resource.api.sap/v1",
  "data": {
```

**Comment by howardjohn** [comment, marked as answer]
Is this selected as TCP protocol for resourcemanager service? https://istio.io/latest/docs/ops/configuration/traffic-management/protocol-selection/
